packages/@lwc/engine-core/src/framework/vm.ts

```diff
@@ -257,6 +261,16 @@ export function createVM<HostNode, HostElement>(
 ): VM {
     const { mode, owner, renderer, tagName } = options;
 
+    let shadowMode;
+    if (renderer.syntheticShadow) {
+        shadowMode =
+            def.preferNativeShadow && isNativeShadowRootDefined
+                ? ShadowMode.Native
+                : ShadowMode.Synthetic;
```

Comment:
This code can be moved in `def.ts` in `createComponentDef`. When a component is registered we can compute the shadow mode and store it in the `ComponentDef`.

Reply:
Curious what your thoughts are regarding this thread:
https://github.com/salesforce/lwc/pull/2358#discussion_r646538644

Reply:
@pmdartus I don't think that possible if we will have the transitivity mode enabled. Basically, this value should depend on the value of the owner vm, otherwise you might ended up with native -> synthetic child, which will not work.

Reply:
@ekashida I think that part of the logic is missing here, when the vm is created, if the ownerVM is using native, the new vm MUST use native or throw a big error, and I hope that we can just force the native and hope for the best.

Reply:
>  if the ownerVM is using native, the new vm MUST use native or throw a big error, and I hope that we can just force the native and hope for the best

@nolanlawson is working on this in a parallel PR: https://github.com/salesforce/lwc/pull/2360

> Basically, this value should depend on the value of the owner vm, otherwise you might ended up with native -> synthetic child, which will not work.

I don't think the `shadowMode` should depend on the owner value. As you pointed out in your previous comment, attempting to create a synthetic shadow component inside a native shadow component should throw an error and fail the VM creation. 

That said the actual `shadowMode` value is not influenced by the owner `shadowMode`.

Reply:
Where can I read more about that decision?

Reply:
Adding this topic to the next dev sync.